payments-ui-core/src/test/java/com/stripe/android/ui/core/elements/CardNumberControllerTest.kt

```diff
@@ -44,66 +44,56 @@ internal class CardNumberControllerTest {
     }
 
     @Test
-    fun `When invalid card number verify visible error`() {
-        val errorFlowValues = mutableListOf<FieldError?>()
-        cardNumberController.error.asLiveData()
-            .observeForever {
-                errorFlowValues.add(it)
-            }
+    fun `When invalid card number verify visible error`() = runTest {
+        cardNumberController.error.distinctUntilChanged().test {
+            assertThat(awaitItem()).isNull()
 
-        cardNumberController.onValueChange("012")
-        idleLooper()
+            cardNumberController.onValueChange("012")
+            idleLooper()
 
-        assertThat(errorFlowValues[errorFlowValues.size - 1]?.errorMessage)
-            .isEqualTo(StripeR.string.stripe_invalid_card_number)
+            skipItems(2)
+
+            assertThat(awaitItem()?.errorMessage)
+                .isEqualTo(StripeR.string.stripe_invalid_card_number)
+        }
     }
 
     @Test
-    fun `Verify get the form field value correctly`() {
-        val formFieldValuesFlow = mutableListOf<FormFieldEntry?>()
-        cardNumberController.formFieldValue.asLiveData()
-            .observeForever {
-                formFieldValuesFlow.add(it)
-            }
-
-        cardNumberController.onValueChange("4242")
-        idleLooper()
-
-        assertThat(formFieldValuesFlow[formFieldValuesFlow.size - 1]?.isComplete)
-            .isFalse()
-        assertThat(formFieldValuesFlow[formFieldValuesFlow.size - 1]?.value)
-            .isEqualTo("4242")
-
-        cardNumberController.onValueChange("4242424242424242")
-        idleLooper()
-
-        assertThat(formFieldValuesFlow[formFieldValuesFlow.size - 1]?.isComplete)
-            .isTrue()
-        assertThat(formFieldValuesFlow[formFieldValuesFlow.size - 1]?.value)
-            .isEqualTo("4242424242424242")
+    fun `Verify get the form field value correctly`() = runTest {
+        cardNumberController.formFieldValue.distinctUntilChanged().test {
```

Comment:
Should we use `distinctUntilChanged()` here? Is this different from the production behavior?

Reply:
The `distinctUntilChanged` just makes it easier to test. Production code doesn't do a `distinctUntilChanged`, although it probably could. But my goal was to change no production code for this change.

Reply:
Let’s merge like this then and add `distinctUntilChanged` in production code in a follow-up?